tests/v1/kv_connector/unit/test_invalid_blocks_race_conditions.py

```diff
@@ -0,0 +1,542 @@
+
+    num_prompt_tokens = num_prompt_blocks * fail_scheduler.block_size
+    num_external_computed_tokens = (
+        num_external_computed_blocks * fail_scheduler.block_size
+    )
+
+    request = create_request(num_tokens=num_prompt_tokens)
+    fail_scheduler.add_request(request=request)
+
+    req_num_new_matched_tokens = {
+        request.request_id: num_external_computed_tokens,
+    }
+
+    # mock connector indicating async load
+    fail_scheduler.connector = Mock()
+    fail_scheduler.connector.get_num_new_matched_tokens.side_effect = (
+        _make_get_num_new_matched_tokens(req_num_new_matched_tokens, True)
+    )
+
+    # critical: request_finished returns delay_free_blocks=True for async transfers
+    # this prevents blocks from being freed while transfer is in progress
+    fail_scheduler.connector.request_finished.return_value = (True, None)
+    fail_scheduler.connector.take_events.return_value = ()
+
+    scheduler_output = fail_scheduler.schedule()
+
+    # request should be waiting for remote KVs
+    assert len(fail_scheduler.waiting) == 1
+    assert request.status == RequestStatus.WAITING_FOR_REMOTE_KVS
+    assert request.num_computed_tokens == 0
+
+    # get the allocated block IDs
+    (req_block_ids,) = fail_scheduler.kv_cache_manager.get_block_ids(request.request_id)
+    invalid_block_ids = {req_block_ids[invalid_block_idx]}
+
+    # report invalid blocks (transfer not finished yet)
+    model_runner_output = create_model_runner_output(
+        reqs=[],
+        finished_recving=None,  # transfer NOT finished
+        invalid_block_ids=invalid_block_ids,
+        use_eos=True,
+    )
+
+    outputs = fail_scheduler.update_from_output(scheduler_output, model_runner_output)
+
+    # request should be marked as finished with error
+    assert request.status == RequestStatus.FINISHED_ERROR
+    assert request.get_finished_reason() == FinishReason.ERROR
+
+    # output should be generated
+    assert len(outputs) == 1
+    engine_outputs = next(iter(outputs.values()))
+    assert len(engine_outputs.outputs) == 1
+    output = engine_outputs.outputs[0]
+    assert output.request_id == request.request_id
+    assert output.finish_reason == FinishReason.ERROR
+
+    # critical: because delay_free_blocks=True, blocks should NOT be freed yet
+    assert request.request_id in fail_scheduler.requests, (
+        "Request should remain in scheduler.requests when delay_free_blocks=True"
+    )
+
+    # verify blocks are still allocated and protected
+    allocated_blocks = fail_scheduler.kv_cache_manager.get_block_ids(request.request_id)
+    assert allocated_blocks is not None
+    assert len(allocated_blocks[0]) > 0, (
+        "Blocks should be protected from freeing when delay_free_blocks=True"
+    )
+
+
+def test_sync_recompute_blocks_not_freed_for_running_requests(
+    recompute_scheduler: Scheduler,
+):
+    """
+    Test Issue 2: Sync recompute case - blocks incorrectly freed for running requests.
```

Comment:
I wouldn't call this a "race condition"? (as in, not sure about the filename choice)